Sources/OpenCombine/Publishers/Publishers.Zip.swift

```diff
@@ -0,0 +1,697 @@
+    }
+}
+
+extension InnerBase: Subscription {
+    fileprivate final func request(_ demand: Subscribers.Demand) {
+        guard demand != .none else {
+            fatalError()
+        }
+        lock.lock()
+        downstreamDemand += demand
+        sendRequestUpstream(demand: demand)
+        if valueIsBeingProcessed {
+            demandReceivedWhileProcessing = demand
+        } else {
+            valueIsBeingProcessed = true
+            processValue()
+        }
+        lock.unlock()
+    }
+
+    fileprivate final func cancel() {
+        lock.lock()
+        let subscriptionsToCancel = upstreamSubscriptions
+        lock.unlock()
+        subscriptionsToCancel.forEach { $0.cancel() }
+    }
+}
+
+private enum ChildState {
+    case waitingForSubscription
+    case active
+    case finished
+    case failed
+    case canceled
+}
+
+// Note that it's critical that this protocol not have any associated types - specifically
+// note that it does not refer to `Upstream`.
+// This allows `InnerBase` to do most of the heavy lifting without regard to the
+// upstream publisher's value type.
+private protocol ChildSubscription: AnyObject, Subscription {
+    var state: ChildState { get set }
+    var childIndex: Int { get }
+    var hasValue: Bool { get }
+}
+
+fileprivate final class ChildSubscriber<Upstream: Publisher, Downstream: Subscriber>
+    where Upstream.Failure == Downstream.Failure
+{
+    typealias Input = Upstream.Output
+    typealias Failure = Upstream.Failure
+
+    fileprivate final var state: ChildState = .waitingForSubscription
+    fileprivate final var upstreamSubscription: Subscription?
+    private var values = [Upstream.Output]()
+    private unowned let parent: InnerBase<Downstream>
+    fileprivate let childIndex: Int
+
+    init(_ parent: InnerBase<Downstream>, _ childIndex: Int) {
+        self.parent = parent
+        self.childIndex = childIndex
+    }
+
+    fileprivate final func dequeueValue() -> Upstream.Output {
+        return values.remove(at: 0)
+    }
+}
+
+extension ChildSubscriber: ChildSubscription {
+    fileprivate final var hasValue: Bool {
+        return !values.isEmpty
+    }
+}
+
+extension ChildSubscriber: Subscription {
+    fileprivate final func request(_ demand: Subscribers.Demand) {
+        upstreamSubscription?.request(demand)
+    }
+}
+
+extension ChildSubscriber: Cancellable {
+    fileprivate final func cancel() {
+        upstreamSubscription?.cancel()
+        upstreamSubscription = nil
+    }
+}
+
+extension ChildSubscriber: Subscriber {
+    fileprivate final func receive(subscription: Subscription) {
+        if upstreamSubscription == nil {
+            upstreamSubscription = subscription
+            parent.receivedSubscription(for: self)
+        } else {
+            subscription.cancel()
+        }
+    }
+
+    fileprivate final func receive(_ input: Input) -> Subscribers.Demand {
+        return parent.receivedChildValue(child: self) { values.append(input) }
```

Comment:
Hi, I am having a crash when using Zip2.

The reason seems to be the unowned parent, when `receive(_ input: Input)` gets called the parent was deallocated and the app crashes.
```swift
fileprivate final func receive(_ input: Input) -> Subscribers.Demand {
    return parent.receivedChildValue(child: self) { values.append(input) }
}
```
Using weak fixes the problem(because of the circular dependency between Publishers.Zip.Inner and ChildSubscriber)
Snipet with the changes
```swift
fileprivate final class ChildSubscriber<Upstream: Publisher, Downstream: Subscriber>
    where Upstream.Failure == Downstream.Failure
{
    private weak var parent: InnerBase<Downstream>?
}

extension ChildSubscriber: Subscriber {
    fileprivate final func receive(subscription: Subscription) {
        if upstreamSubscription == nil, let parent = parent {
            upstreamSubscription = subscription
            parent.receivedSubscription(for: self)
        } else {
            subscription.cancel()
        }
    }

    fileprivate final func receive(_ input: Input) -> Subscribers.Demand {
        guard let parent = parent else { return .none }
        return parent.receivedChildValue(child: self) { values.append(input) }
    }

    fileprivate final func receive(completion: Subscribers.Completion<Failure>) {
        guard let parent = parent else { return }
        parent.receivedCompletion(completion, forChild: self)
    }
}
```
Could you take a look?
I was trying to build an isolated failing test but I have not been able to reproduce it in a controlled environment.

Thanks in advance :)